src/MLJInterface.jl

```diff
@@ -0,0 +1,309 @@
+
+# TODO: Enable `verbosity` being passed to `equation_search`
+function MMI.fit(m::AbstractSRRegressor, verbosity, X, y, w=nothing)
+    return MMI.update(m, verbosity, (; state=nothing), nothing, X, y, w)
+end
+function MMI.update(
+    m::AbstractSRRegressor, verbosity, old_fitresult, old_cache, X, y, w=nothing
+)
+    options = get(old_fitresult, :options, get_options(m))
+    X_t, variable_names = get_matrix_and_colnames(X)
+    y_t = format_input_for(m, y)
+    search_state = equation_search(
+        X_t,
+        y_t;
+        niterations=m.niterations,
+        weights=w,
+        variable_names=variable_names,
+        options=options,
+        parallelism=m.parallelism,
+        numprocs=m.numprocs,
+        procs=m.procs,
+        addprocs_function=m.addprocs_function,
+        runtests=m.runtests,
+        saved_state=old_fitresult.state,
+        return_state=true,
+        loss_type=m.loss_type,
+    )
+    fitresult = (; state=search_state, options=options, variable_names=variable_names)
+    return (fitresult, nothing, full_report(m, fitresult))
+end
+function get_matrix_and_colnames(X)
+    sch = MMI.istable(X) ? MMI.schema(X) : nothing
+    Xm_t = MMI.matrix(X; transpose=true)
+    colnames = if sch === nothing
+        [map(i -> "x$(i)", axes(Xm_t, 1))...]
+    else
+        [string.(sch.names)...]
+    end
+    return Xm_t, colnames
+end
+
+function format_input_for(::SRRegressor, y)
+    @assert(
+        !(MMI.istable(y) || (length(size(y)) == 2 && size(y, 2) > 1)),
+        "For multi-output regression, please use `MultitargetSRRegressor`."
+    )
+    return vec(y)
+end
+function format_input_for(::MultitargetSRRegressor, y)
+    @assert(
+        MMI.istable(y) || (length(size(y)) == 2 && size(y, 2) > 1),
+        "For single-output regression, please use `SRRegressor`."
+    )
+    return MMI.matrix(y; transpose=true)
+end
+
+function MMI.fitted_params(m::AbstractSRRegressor, fitresult)
+    report = full_report(m, fitresult)
+    return (;
+        best_idx=report.best_idx,
+        equations=report.equations,
+        equation_strings=report.equation_strings,
+    )
+end
+function MMI.predict(m::SRRegressor, fitresult, Xnew)
+    params = MMI.fitted_params(m, fitresult)
+    Xnew_t = MMI.matrix(Xnew; transpose=true)
+    eq = params.equations[params.best_idx]
+    out, flag = eval_tree_array(eq, Xnew_t, fitresult.options)
+    !flag && error("Detected a NaN in evaluating expression.")
+    return out
+end
+function MMI.predict(m::MultitargetSRRegressor, fitresult, Xnew)
+    params = MMI.fitted_params(m, fitresult)
+    Xnew_t = MMI.matrix(Xnew; transpose=true)
+    equations = params.equations
+    best_idx = params.best_idx
+    outs = [
+        let (out, flag) = eval_tree_array(eq[i], Xnew_t, fitresult.options)
+            !flag && error("Detected a NaN in evaluating expression.")
+            out
+        end for (i, eq) in zip(best_idx, equations)
+    ]
+    return reduce(hcat, outs)
```

Comment:
```suggestion
    outs_matrix = reduce(hcat, outs)
    prototype = MMI.istable(Xnew) ? Xnew : nothing
    return MMI.table(outs_matrix, prototype=prototype)
```

Reply:
predictions from multitarget regressor type models must be tables

Reply:
I'm trying this now. I think this assumes `X` has the same shape and variable names as `y`, no? 

Also, what if the user passes `X` as a regular matrix? Then, e.g.,

```julia
predict(mach, X) .- y
```

would throw an error, since `predict(mach, X)` is a table and `y` is a matrix.

Reply:
Can we store the schema of both `y` and `X` in `fitresult`, and use it as the `prototype`?

Reply:
> I'm trying this now. I think this assumes `X` has the same shape and variable names as `y`, no?
> 
> Also, what if the user passes `X` as a regular matrix? Then, e.g.,
> 
> ```julia
> predict(mach, X) .- y
> ```
> 
> would throw an error, since `predict(mach, X)` is a table and `y` is a matrix.

We can work around that case

Reply:
Check out what I just implemented. I think it solves both cases:

```julia
    X_t, variable_names, schema = get_matrix_and_colnames(X)
    y_t, y_variable_names, y_schema = format_input_for(m, y)
```

to get the column names and schema of both `y` and `X` (if necessary). Then, these are used in `predict` to give back a table of the same schema:

```julia
    outs = [
        let (out, flag) = eval_tree_array(eq[i], Xnew_t, fitresult.options)
            !flag && error("Detected a NaN in evaluating expression.")
            out
        end for (i, eq) in zip(best_idx, equations)
    ]
    outs_matrix = reduce(hcat, outs)
    fitresult.y_schema === nothing && return outs_matrix
    return MMI.table(
        outs_matrix; names=fitresult.y_variable_names, prototype=fitresult.y_schema
    )
```
This gives you a table if you train on a table for `y`, and a matrix if you train on a matrix.

Reply:
prototype accepts a table source

Reply:
We have to check in stages.

Reply:
> prototype accepts a table source not the schema

Should we store the original `y` input in `fitresult` so that we can use the correct prototype? Or are there any other lightweight things we could store?